File: src/518.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    int change(int amount, vector<int>& coins) {
        int retVal = 0;

        vector<int> dp(amount + 1, 0);
        dp[0] = 1;
        int coinsSize = coins.size();
        for (int i = 0; i < coinsSize; ++i) {
            for (int j = coins[i]; j <= amount; ++j) {
                dp[j] += dp[j - coins[i]];
            }
        }
        retVal = dp[amount];

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        int amount;
        vector<int> coins;
    };
    vector<subject> testData{{5, {1, 2, 5}}, {3, {2}}, {10, {10}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: amount = 5, coins = [1,2,5]
     *  Output: 4
     *
     *  Input: amount = 3, coins = [2]
     *  Output: 0
     *
     *  Input: amount = 10, coins = [10]
     *  Output: 1
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: amount = " << testData[i].amount << ", coins = [";
        for (long unsigned int j = 0; j < testData[i].coins.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].coins[j];
        }
        cout << "]\n";

        answer = cSolution.change(testData[i].amount, testData[i].coins);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/518.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        retVal = 0

        dp = [0] * (amount+1)
        dp[0] = 1
        coinsSize = len(coins)
        for i in range(coinsSize):
            for j in range(coins[i], amount+1):
                dp[j] += dp[j - coins[i]]
        retVal = dp[amount]

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for amount, coins in zip([5, 3, 10], [[1, 2, 5], [2], [10]]):
            # /* Example
            #  *  Input: amount = 5, coins = [1,2,5]
            #  *  Output: 4
            #  *
            #  *  Input: amount = 3, coins = [2]
            #  *  Output: 0
            #  *
            #  *  Input: amount = 10, coins = [10]
            #  *  Output: 1
            #  */
            logging.info("Input: amount = %s, coins = %s", amount, coins)

            retVal = pSolution.change(amount, coins)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
